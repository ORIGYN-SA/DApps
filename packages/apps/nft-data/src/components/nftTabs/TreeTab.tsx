import React, { useEffect } from 'react';
import { Box, FormControlLabel, Switch } from '@mui/material';
import TreeView from '@mui/lab/TreeView';
import TreeItem from '@mui/lab/TreeItem';
import pick from 'lodash/pick';
import ExpandMoreIcon from '@mui/icons-material/ExpandMore';
import ChevronRightIcon from '@mui/icons-material/ChevronRight';
import { useState } from 'react';
import { Container } from '@origyn-sa/origyn-art-ui';
interface RenderTree {
  id: string;
  name: string;
  children?: readonly RenderTree[];
}

function Tree({ metadata }: any) {
  let all_index = 0;
  let apps_index: number[] = [0, 0];
  let libraries_index: number[] = [0, 0];
  const [expanded, setExpanded] = useState<string[]>([]);
  const [all, setAll] = React.useState(true);
  const [apps, setApps] = React.useState(false);
  const [libraries, setLibraries] = React.useState(false);

  const init = (start, end) => {
    let array: [string] = ['0'];
    array.push(start + '');
    if (Object.entries(metadata).length) {
      for (let i = start; i <= end; i++) {
        array.push(`${i}`);
      }
    }
    return array;
  };

  const handleExpandAll = (event: React.ChangeEvent<HTMLInputElement>) => {
    setAll(event.target.checked);
    if (event.target.checked) {
      setExpanded([...init(0, all_index)]);
    } else {
      setExpanded([]);
    }
    setApps(false);
    setLibraries(false);
  };

  const handleExpandApps = (event: React.ChangeEvent<HTMLInputElement>) => {
    setApps(event.target.checked);
    if (event.target.checked) {
      setExpanded([...init(apps_index[0], apps_index[1])]);
    } else {
      setExpanded([]);
    }
    setAll(false);
    setLibraries(false);
  };

  const handleExpandLibraries = (event: React.ChangeEvent<HTMLInputElement>) => {
    setLibraries(event.target.checked);
    if (event.target.checked) {
      setExpanded([...init(libraries_index[0], libraries_index[1])]);
    } else {
      setExpanded([]);
    }
    setAll(false);
    setApps(false);
  };

  const increment = () => {
    all_index++;
    return `${all_index}`;
  };
  const handleToggle = (event, nodeIds) => {
    setExpanded(nodeIds);
  };

  useEffect(() => {
    setExpanded([...init(0, all_index)]);
  }, [metadata]);

  const parseData = (data) => {
    let arr: RenderTree[] = [];
    let obj = pick(data, [
      'owner',
      'hidden_asset',
      'preview_asset',
      'primary_asset',
      'experience_asset',
      'id',
    ]);

    for (let i in obj) {
      arr.push({ id: increment(), name: i, children: [{ id: increment(), name: obj[i] }] });
    }

    arr.push({
      id: increment(),
      name: '__apps',
      children: data.__apps?.map((item, i) => {
        apps_index[0] = all_index;

        let obj1 = { ...item.data };
        let arr1: RenderTree[] = [];
        for (let j in obj1) {
          arr1.push({ id: increment(), name: j, children: [{ id: increment(), name: obj1[j] }] });
        }

        return {
          id: increment(),
          name: `${i + 1}`,
          children: [
            { id: increment(), name: 'app_id', children: [{ id: increment(), name: item.app_id }] },
            { id: increment(), name: 'read', children: [{ id: increment(), name: item.read }] },
            {
              id: increment(),
              name: 'write',
              children: [
                {
                  id: increment(),
                  name: 'type',
                  children: [{ id: increment(), name: item.write.type }],
                },
                {
                  id: increment(),
                  name: 'list',
                  children: item.write.list.map((x) => ({ id: increment(), name: x })),
                },
              ],
            },
            {
              id: increment(),
              name: 'permissions',
              children: [
                {
                  id: increment(),
                  name: 'type',
                  children: [{ id: increment(), name: item.permissions.type }],
                },
                {
                  id: increment(),
                  name: 'list',
                  children: item.permissions.list.map((x) => ({ id: increment(), name: x })),
                },
              ],
            },
            { id: increment(), name: 'data', children: [...arr1] },
          ],
        };
      }),
    });
    apps_index[1] = all_index;

    libraries_index[0] = all_index + 1;
    arr.push({
      id: increment(),
      name: 'library',
      children: data.library?.map((item, i) => {
        let obj1: RenderTree[] = [];
        for (let j in item) {
          obj1.push({ id: increment(), name: j, children: [{ id: increment(), name: item[j] }] });
        }

        return { id: increment(), name: i + 1, children: [...obj1] };
      }),
    });
    libraries_index[1] = all_index;

    return arr;
  };
  const data: RenderTree = {
    id: '0',
    name: 'NFT',
    children: [...parseData(metadata)],
  };

  const renderTree = (nodes: RenderTree) => (
    <TreeItem key={nodes.id} nodeId={nodes.id} label={nodes.name}>
      {Array.isArray(nodes.children) ? nodes.children.map((node) => renderTree(node)) : null}
    </TreeItem>
  );
  return (
    <Container padding="8px 16px">
    <Box margin="0 0 0 0">
      <FormControlLabel
        control={<Switch onChange={handleExpandAll} checked={all} />}
        label="Expand All"
      />
      <FormControlLabel
        control={<Switch onChange={handleExpandApps} checked={apps} />}
        label="Expand Apps"
      />
      <FormControlLabel
        control={<Switch onChange={handleExpandLibraries} checked={libraries} />}
        label="Expand Libraries"
      />
      {Object.entries(metadata).length > 0 && (
        <TreeView
          aria-label="file system navigator"
          expanded={expanded}
          defaultCollapseIcon={<ExpandMoreIcon />}
          defaultExpandIcon={<ChevronRightIcon />}
          onNodeToggle={handleToggle}
          sx={{ height: 800, flexGrow: 1, maxWidth: 800, overflowY: 'auto' }}
        >
          {renderTree(data)}
        </TreeView>
      )}
    </Box>
    </Container>
  );
}

export default Tree;
