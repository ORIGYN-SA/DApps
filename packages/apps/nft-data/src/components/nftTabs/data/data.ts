import { useContext } from 'react';
import { useRoute, AuthContext } from '@dapp/features-authentication';
import { getNft, getNftCollectionMeta, OrigynClient } from '@origyn/mintjs';

export type Nft_Data = {
  name: string;
  value: string | number;
  immutable: boolean;
  level: Level;
};

export type Level = 'Asset' | 'App' | 'Id' | 'Permission';

export type Permission = {
  level: Level;
  type: 'Permission' | 'Write' | 'Read';
  name: string;
  value?: string;
  list?: any[];
  immutable: boolean;
};

const getMetadata = async () => {
  const { actor } = useContext(AuthContext);
  const { tokenId, canisterId } = await useRoute();
  await OrigynClient.getInstance().init(true, canisterId, { actor });
  if (tokenId) {
    const response = await getNft(tokenId);
    console.log('responseNFT', response);
    const metadata = await response.ok.metadata.Class;
    console.log('metadataNft', metadata);
    return metadata;
  } else {
    console.log('no token id - getNftCollectionMeta');
    const response = await getNftCollectionMeta();
    console.log('responseCollection', response);
    const metadata = await response.ok.metadata[0].Class;
    console.log('collectionMeta', metadata);
    return metadata;
  }
};

export const getData = async () => {
  const { tokenId } = await useRoute();
  let Data_Array: Nft_Data[] = [];
  const Metadata = await getMetadata();
  console.log('Metadata', Metadata);
  if (Metadata) {
    try {
      Data_Array.push(
        {
          name: 'Id',
          value: await Metadata.filter((res) => {
            return res.name === 'id';
          })[0].value.Text,
          immutable: await Metadata.filter((res) => {
            return res.name === 'id';
          })[0].immutable,
          level: 'Id',
        },
        {
          name: 'Owner',
          value: await Metadata.filter((res) => {
            return res.name === 'owner';
          })[0].value.Text,
          immutable: await Metadata.filter((res) => {
            return res.name === 'owner';
          })[0].immutable,
          level: 'Id',
        },
        {
          name: 'App_Id',
          value: await Metadata.filter((res) => {
            return res.name === '__apps';
          })[0].value.Array.thawed[0].Class[0].value.Text,
          immutable: await Metadata.filter((res) => {
            return res.name === '__apps';
          })[0].value.Array.thawed[0].Class[0].immutable,
          level: 'App',
        },
        {
          name: 'Read',
          value: await Metadata.filter((res) => {
            return res.name === '__apps';
          })[0].value.Array.thawed[0].Class[1].value.Text,
          immutable: await Metadata.filter((res) => {
            return res.name === '__apps';
          })[0].value.Array.thawed[0].Class[1].immutable,
          level: 'Permission',
        },
      );
      // Assets level - Only for NFTs, Not for collecrion
      if (tokenId) {
        Data_Array.push(
          {
            name: 'Preview',
            value: await Metadata.filter((res) => {
              return res.name === 'preview_asset';
            })[0].value.Text,
            immutable: await Metadata.filter((res) => {
              return res.name === 'preview_asset';
            })[0].immutable,
            level: 'Asset',
          },
          {
            name: 'Hidden',
            value: await Metadata.filter((res) => {
              return res.name === 'hidden_asset';
            })[0].value.Text,
            immutable: await Metadata.filter((res) => {
              return res.name === 'hidden_asset';
            })[0].immutable,
            level: 'Asset',
          },
          {
            name: 'Primary',
            value: await Metadata.filter((res) => {
              return res.name === 'primary_asset';
            })[0].value.Text,
            immutable: await Metadata.filter((res) => {
              return res.name === 'primary_asset';
            })[0].immutable,
            level: 'Asset',
          },
          {
            name: 'Experience',
            value: await Metadata.filter((res) => {
              return res.name === 'experience_asset';
            })[0].value.Text,
            immutable: await Metadata.filter((res) => {
              return res.name === 'experience_asset';
            })[0].immutable,
            level: 'Asset',
          },
        );
      }
      // Apps level
      const apps = await Metadata.filter((res) => {
        return res.name === '__apps';
      })[0].value.Array.thawed[0].Class[4].value.Class;
      let i: string;
      for (i in apps) {
        console.log(apps[i]);
        if (apps[i].value.hasOwnProperty('Principal')) {
          Data_Array.push({
            name: apps[i].name,
            value: apps[i].value.Principal.toText(),
            immutable: apps[i].immutable,
            level: 'App',
          });
        } else if (apps[i].value.hasOwnProperty('Nat')) {
          Data_Array.push({
            name: apps[i].name,
            value: apps[i].value.Nat.toString(),
            immutable: apps[i].immutable,
            level: 'App',
          });
        } else {
          Data_Array.push({
            name: apps[i].name,
            value: apps[i].value.Text,
            immutable: apps[i].immutable,
            level: 'App',
          });
        }
      }
    } catch (e) {
      console.log(e);
    }
  }
  console.log('Data', Data_Array);
  return Data_Array;
};

export const getPermissions = async () => {
  let Allowed_Array: Permission[] = [];
  const Metadata = await getMetadata();

  if (Metadata) {
    try {
      let array_write = await Metadata.filter((res) => {
        return res.name === '__apps';
      })[0].value.Array.thawed[0].Class[2].value.Class;
      let i: any;
      for (i in array_write) {
        if (array_write[i].value.hasOwnProperty('Array')) {
          let list = [];
          let j: any;
          for (j in array_write[i].value.Array.thawed) {
            list.push(array_write[i].value.Array.thawed[j].Principal.toText());
          }

          Allowed_Array.push({
            level: 'Permission',
            name: array_write[i].name,
            immutable: array_write[i].immutable,
            list: list,
            type: 'Write',
          });
        } else {
          Allowed_Array.push({
            level: 'Permission',
            name: array_write[i].name,
            immutable: array_write[i].immutable,
            value: array_write[i].value.Text,
            type: 'Write',
          });
        }
      }

      let array_permission = await Metadata.filter((res) => {
        return res.name === '__apps';
      })[0].value.Array.thawed[0].Class[3].value.Class;
      let k: any;
      for (k in array_permission) {
        if (array_permission[k].value.hasOwnProperty('Array')) {
          let list = [];
          let l: any;
          for (l in array_permission[k].value.Array.thawed) {
            list.push(array_permission[k].value.Array.thawed[l].Principal.toText());
          }

          Allowed_Array.push({
            level: 'Permission',
            name: array_permission[k].name,
            immutable: array_permission[k].immutable,
            list: list,
            type: 'Permission',
          });
        } else {
          Allowed_Array.push({
            level: 'Permission',
            name: array_permission[k].name,
            immutable: array_permission[k].immutable,
            value: array_permission[k].value.Text,
            type: 'Permission',
          });
        }
      }
    } catch (e) {
      console.log(e);
    }
  }
  return Allowed_Array;
};
