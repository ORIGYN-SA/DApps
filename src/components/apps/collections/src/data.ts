// src/data.ts

interface Collection {
  name: string;
  checked: boolean;
  image: string;
  category_id: string;
  nftCount: number;
}

interface BackendResponse {
  collections: Collection[];
  totalPages: number;
}

const fetchCollectionsFromBackend = async (
  page: number,
  limit: number,
): Promise<BackendResponse> => {
  // Simulate more data for pagination
  const allCollections: Collection[] = [
    {
      name: 'Art',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Alice',
      nftCount: 120,
    },
    {
      name: 'Precious Metal',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Bob',
      nftCount: 85,
    },
    {
      name: 'Diamond',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Charlie',
      nftCount: 200,
    },
    {
      name: 'Jewelry',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Diana',
      nftCount: 150,
    },
    {
      name: 'Music',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Ethan',
      nftCount: 95,
    },
    {
      name: 'Real Estate',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Fiona',
      nftCount: 60,
    },
    {
      name: 'Photography',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Grace',
      nftCount: 110,
    },
    {
      name: 'Vintage Cars',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Henry',
      nftCount: 75,
    },
    {
      name: 'Antiques',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Irene',
      nftCount: 50,
    },
    {
      name: 'Collectible Coins',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Jack',
      nftCount: 130,
    },
    // Add more collections to have enough data for pagination
    {
      name: 'Sculptures',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Leo',
      nftCount: 90,
    },
    {
      name: 'Books',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Mia',
      nftCount: 65,
    },
    {
      name: 'Virtual Real Estate',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Nolan',
      nftCount: 80,
    },
    {
      name: 'Sports Memorabilia',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Olivia',
      nftCount: 115,
    },
    {
      name: 'Watches',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Peter',
      nftCount: 55,
    },
    {
      name: 'Wine Collection',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Quincy',
      nftCount: 70,
    },
    {
      name: 'Fashion',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Rachel',
      nftCount: 125,
    },
    {
      name: 'Handicrafts',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Sam',
      nftCount: 100,
    },
    {
      name: 'Rare Stamps',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Tina',
      nftCount: 40,
    },
    {
      name: 'Ceramics',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Uma',
      nftCount: 85,
    },
    {
      name: 'Modern Art',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Victor',
      nftCount: 95,
    },
    {
      name: 'Classic Cars',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Wendy',
      nftCount: 60,
    },
    {
      name: 'Historical Documents',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Xavier',
      nftCount: 30,
    },
    {
      name: 'Comics',
      checked: true,
      image: 'https://via.placeholder.com/150',
      category_id: 'Zach',
      nftCount: 110,
    },
  ];

  const startIndex = (page - 1) * limit;
  const paginatedCollections = allCollections.slice(startIndex, startIndex + limit);
  const totalPages = Math.ceil(allCollections.length / limit);

  return {
    collections: paginatedCollections,
    totalPages,
  };
};
