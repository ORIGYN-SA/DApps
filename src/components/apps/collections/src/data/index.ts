import { Actor, HttpAgent } from '@dfinity/agent';
import { idlFactory } from './canisters/gold/did';
import { Principal } from '@dfinity/principal';
import { _SERVICE } from '../data/canisters/gold/interfaces/gld_nft';

export interface Collection {
  name: string;
  checked: boolean;
  image: string;
  creator: string;
  nftCount: number;
}

export interface BackendResponse {
  collections: Collection[];
  totalPages: number;
}

export const fetchCollectionsFromBackend = async (
  page: number,
  itemsPerPage: number,
): Promise<BackendResponse> => {
  try {
    const agent = new HttpAgent({ host: 'https://ic0.app' });
    const canisterId = 'io7gn-vyaaa-aaaak-qcbiq-cai';
    const actor = Actor.createActor<_SERVICE>(idlFactory, {
      agent,
      canisterId,
    });

    const logoResult = await actor.icrc7_logo();
    console.log('logoResult', logoResult);

    const collectionResult = await actor.collection_nft_origyn([]);
    if (!('ok' in collectionResult)) {
      throw new Error(
        `Erreur lors de la récupération de la collection : ${collectionResult.err.text}`,
      );
    }
    const collectionInfo = collectionResult.ok;

    if (!collectionInfo.token_ids || collectionInfo.token_ids.length === 0) {
      return { collections: [], totalPages: 0 };
    }

    const tokenIds = collectionInfo.token_ids[0];

    const nftResults = await actor.nft_batch_origyn(tokenIds);

    const collections: Collection[] = nftResults
      .map((nftResult, index) => {
        if ('ok' in nftResult) {
          const metadata = nftResult.ok.metadata;
          const name = extractName(metadata);
          const creator = extractCreator(metadata);
          const nftCount = 1; // TODO
          return {
            name,
            checked: true,
            image: logoResult ? logoResult : '',
            creator,
            nftCount,
          };
        } else {
          console.error(`Erreur pour le token ${tokenIds[index]}: ${nftResult.err.text}`);
          return undefined;
        }
      })
      .filter((item): item is Collection => item !== undefined);

    const totalPages = Math.ceil(collections.length / itemsPerPage);

    console.log('collections', collections);

    return {
      collections,
      totalPages,
    };
  } catch (error) {
    console.error('Erreur dans fetchCollectionsFromBackend:', error);
    return { collections: [], totalPages: 0 };
  }
};

const extractName = (metadata: any): string => {
  if ('Class' in metadata) {
    const properties = metadata.Class;

    const appsProp = properties.find((prop: any) => prop.name === '__apps');
    if (appsProp && 'Array' in appsProp.value) {
      const appsArray = appsProp.value.Array;

      for (const app of appsArray) {
        if ('Class' in app) {
          const appProps = app.Class;

          const appIdProp = appProps.find(
            (prop: any) =>
              prop.name === 'app_id' && prop.value.Text === 'com.origyn.metadata.general',
          );

          if (appIdProp) {
            const dataProp = appProps.find(
              (prop: any) => prop.name === 'data' && 'Class' in prop.value,
            );
            if (dataProp) {
              const dataProps = dataProp.value.Class;

              const displayNameProp = dataProps.find(
                (prop: any) => prop.name === 'display_name' && 'Text' in prop.value,
              );

              if (displayNameProp) {
                return displayNameProp.value.Text;
              }
            }
          }
        }
      }
    }
  }
  return 'Nom inconnu';
};

const extractCreator = (metadata: any): string => {
  if ('Class' in metadata) {
    const properties = metadata.Class;

    const ownerProp = properties.find(
      (prop: any) => prop.name === 'owner' && prop.value && prop.value.Principal,
    );

    if (ownerProp) {
      const principalValue = ownerProp.value.Principal;

      if (principalValue && typeof principalValue.toText === 'function') {
        return principalValue.toText();
      } else if (principalValue._arr) {
        const arrUnknown = Object.values(principalValue._arr);

        const arr = arrUnknown.map((item: unknown) => {
          if (typeof item === 'number') {
            return item;
          } else {
            return 0;
          }
        });

        const uint8Array = new Uint8Array(arr);
        const principal = Principal.fromUint8Array(uint8Array);
        return principal.toText();
      }
    }
  }
  return 'Créateur inconnu';
};
