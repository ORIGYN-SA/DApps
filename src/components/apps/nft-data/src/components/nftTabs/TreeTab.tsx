import React from 'react';
import pick from 'lodash/pick';
import { useState } from 'react';
import { Container, Button, Flex } from '@origyn/origyn-art-ui';
import PropTypes from 'prop-types';

interface RenderTree {
  id: string;
  name: string;
  children?: readonly RenderTree[];
}

function Tree({ metadata }: any) {
  let all_index = 0;
  let apps_index: number[] = [0, 0];
  let libraries_index: number[] = [0, 0];
  // const [expanded, setExpanded] = useState<string[]>([]);
  // const [all, setAll] = React.useState(true);
  // const [apps, setApps] = React.useState(false);
  // const [libraries, setLibraries] = React.useState(false);

  // const init = (start, end) => {
  //   let array: [string] = ['0'];
  //   array.push(start + '');
  //   if (Object.entries(metadata).length) {
  //     for (let i = start; i <= end; i++) {
  //       array.push(`${i}`);
  //     }
  //   }
  //   return array;
  // };

  // const handleExpandAll = (event: React.ChangeEvent<HTMLInputElement>) => {
  //   setAll(event.target.checked);
  //   if (event.target.checked) {
  //     setExpanded([...init(0, all_index)]);
  //   } else {
  //     setExpanded([]);
  //   }
  //   setApps(false);
  //   setLibraries(false);
  // };

  // const handleExpandApps = (event: React.ChangeEvent<HTMLInputElement>) => {
  //   setApps(event.target.checked);
  //   if (event.target.checked) {
  //     setExpanded([...init(apps_index[0], apps_index[1])]);
  //   } else {
  //     setExpanded([]);
  //   }
  //   setAll(false);
  //   setLibraries(false);
  // };

  // const handleExpandLibraries = (event: React.ChangeEvent<HTMLInputElement>) => {
  //   setLibraries(event.target.checked);
  //   if (event.target.checked) {
  //     setExpanded([...init(libraries_index[0], libraries_index[1])]);
  //   } else {
  //     setExpanded([]);
  //   }
  //   setAll(false);
  //   setApps(false);
  // };

  const increment = () => {
    all_index++;
    return `${all_index}`;
  };
  // const handleToggle = (event, nodeIds) => {
  //   setExpanded(nodeIds);
  // };

  // useEffect(() => {
  //   setExpanded([...init(0, all_index)]);
  // }, [metadata]);

  const parseData = (data) => {
    let arr: RenderTree[] = [];
    let obj = pick(data, [
      'owner',
      'hidden_asset',
      'preview_asset',
      'primary_asset',
      'experience_asset',
      'id',
    ]);

    for (let i in obj) {
      arr.push({ id: increment(), name: i, children: [{ id: increment(), name: obj[i] }] });
    }
    arr.push({
      id: increment(),
      name: '__system',
      children: [
        {
          id: increment(),
          name: 'status',
          children: [{ id: increment(), name: data.__system.status }],
        },
      ],
    });

    arr.push({
      id: increment(),
      name: '__apps',
      children: data.__apps?.map((item, i) => {
        apps_index[0] = all_index;

        let obj1 = { ...item.data };
        let arr1: RenderTree[] = [];
        for (let j in obj1) {
          arr1.push({
            id: increment(),
            name: j,
            children: [{ id: increment(), name: JSON.stringify(obj1[j]) }],
          });
        }

        return {
          id: increment(),
          name: `${i + 1}`,
          children: [
            { id: increment(), name: 'app_id', children: [{ id: increment(), name: item.app_id }] },
            { id: increment(), name: 'read', children: [{ id: increment(), name: item.read }] },
            {
              id: increment(),
              name: 'write',
              children: [
                {
                  id: increment(),
                  name: 'type',
                  children: [{ id: increment(), name: item.write.type }],
                },
                {
                  id: increment(),
                  name: 'list',
                  children: item.write.list.map((x) => ({ id: increment(), name: x })),
                },
              ],
            },
            {
              id: increment(),
              name: 'permissions',
              children: [
                {
                  id: increment(),
                  name: 'type',
                  children: [{ id: increment(), name: item.permissions.type }],
                },
                {
                  id: increment(),
                  name: 'list',
                  children: item.permissions.list.map((x) => ({ id: increment(), name: x })),
                },
              ],
            },
            { id: increment(), name: 'data', children: [...arr1] },
          ],
        };
      }),
    });
    apps_index[1] = all_index;

    libraries_index[0] = all_index + 1;
    arr.push({
      id: increment(),
      name: 'library',
      children: data.library?.map((item, i) => {
        let obj1: RenderTree[] = [];
        for (let j in item) {
          obj1.push({ id: increment(), name: j, children: [{ id: increment(), name: item[j] }] });
        }

        return { id: increment(), name: i + 1, children: [...obj1] };
      }),
    });
    libraries_index[1] = all_index;

    return arr;
  };
  const data: any = {
    id: '0',
    name: 'NFT',
    children: [...parseData(metadata)],
  };

  const Tree = ({ treeData }) => {
    return (
      <ul>
        {treeData.map((node) => (
          <TreeNode node={node} key={node.id} />
        ))}
      </ul>
    );
  };

  Tree.propTypes = {
    treeData: PropTypes.array,
  };

  const TreeNode = ({ node }) => {
    const { children, name } = node;
    const [showChildren, setShowChildren] = useState(false);
    const handleClick = () => {
      setShowChildren(!showChildren);
    };
    return (
      <>
        <div style={{ marginBottom: '10px' }}>
          {children ? <h5 onClick={handleClick}>{name}</h5> : <span>{name}</span>}
        </div>
        <ul style={{ paddingLeft: '30px' }}>{showChildren && <Tree treeData={children} />}</ul>
      </>
    );
  };

  TreeNode.propTypes = {
    node: PropTypes.object,
  };

  const [showChildren, setShowChildren] = useState(false);

  const handleClick = () => {
    setShowChildren(!showChildren);
  };

  const handleClose = () => {
    setShowChildren(false);
  };

  return (
    <Container padding="16px">
      <Flex flexFlow="row" gap={16}>
        <Button btnType="filled" onClick={handleClose}>
          Close All
        </Button>
        {/*
        <Button btnType="filled" onClick={handleClose}>
          Expand All
        </Button>
        <Button btnType="filled" onClick={handleClose}>
          Expand Libraries
        </Button>
  */}
      </Flex>
      <br />
      <div>
        <h4 onClick={handleClick}>NFT </h4>
        <br />
        {showChildren && <Tree treeData={data.children} />}
      </div>
    </Container>
  );
}

export default Tree;
