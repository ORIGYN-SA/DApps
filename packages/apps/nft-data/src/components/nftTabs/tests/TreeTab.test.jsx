import React from 'react';
import { render, screen, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom';
import TreeTab from '../TreeTab';
import data from './data';
describe('treetab', () => {
    const { asFragment } = render(<TreeTab metadata={data} />);
    it('should return treetab rendered', () => {
        expect(screen.getByText('Expand All')).toBeInTheDocument();
        expect(screen.getByText('Expand Apps')).toBeInTheDocument();
        expect(screen.getByText('Expand Libraries')).toBeInTheDocument();
        expect(screen.getByText('NFT')).toBeInTheDocument();
    });
    it('should match snapshot', () => {
        expect(asFragment()).toMatchSnapshot();
    });

});