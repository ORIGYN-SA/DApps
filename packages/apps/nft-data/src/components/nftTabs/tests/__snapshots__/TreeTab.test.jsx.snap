// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`treetab should match snapshot 1`] = `<DocumentFragment />`;
